### DISCUSSION TITLE
RandomLoadBalance.needWeightLoadBalance judge fail

### DISCUSSION THREAD

**Discussion by TXYan** [apache/dubbo#13874] [General]
- [ ] RandomLoadBalance.needWeightLoadBalance方法获取的是参数invokers的第一个invoker进行判断的，但是可能其他invoker的权重是有设定的，导致应该按权重计算，而使用了随机，导致权重设置没生效

### Environment

* Dubbo version: 3x
* Operating System version: macos
* Java version: 1.8

### Steps to reproduce this issue

1. 启动provider1和provider2
2. 启动一个consumer，封装一个指定次数调用的方法，方法内完成本次provider的ip:port个数统计，循环调用此封装方法
3. 动态改变权重观察每次的统计个数，发现有时候好使，有时候不好使，不好使就是因为第一个invoder是没有权重的，导致走了随机

### Actual Behavior
应该在invoker列表中找任意一个需要权重计算的就需要权重计算

I

**Comment by AlbumenJ** [comment]
这个是性能的取舍，多一次扫描全部的过程对调用的影响就越大，特别是在 provider 数量多的时候

**Comment by TXYan** [comment]
> 这个是性能的取舍，多一次扫描全部的过程对调用的影响就越大，特别是在 provider 数量多的时候

了解，正确性怎么保证，我理解性能应该是正确的基础上进一步优化的，现在我用这个权重调整的功能认为是不好使的

**Comment by TXYan** [comment]
> > 这个是性能的取舍，多一次扫描全部的过程对调用的影响就越大，特别是在 provider 数量多的时候
> 
> 了解，正确性怎么保证，我理解性能应该是正确的基础上进一步优化的，现在我用这个权重调整的功能认为是不好使的

或者是否可以记录计算结果的10秒内存缓存，10秒后生效我觉得还能接受

**Reply by AlbumenJ** [reply]
Cache make it more conplex and we should judge whether we will make it or not.
